File: lib/constants/atributos_categoria.dart
/// Preguntas dinámicas por categoría.
///
/// Espejo de `backend/src/config/atributosCategoria.js`. Está duplicado a
/// propósito y no se descarga: el formulario de publicar tiene que poder
/// pintarse sin esperar a la red, igual que el catálogo de categorías ya
/// tiene su fallback en `mock_data.dart`. Un formulario que aparece medio
/// segundo después de elegir la categoría se siente roto.
///
/// El servidor sigue siendo la autoridad — valida cada respuesta contra SU
/// copia y rechaza lo que no reconoce. Este archivo decide qué se PINTA,
/// no qué se acepta.
///
/// `backend/src/config/atributosCategoria.paridad.test.js` compara ambos
/// archivos y falla si se desincronizan. Si tocas uno, toca el otro.
library;

enum AtributoTipo { booleano, seleccion, seleccionMultiple, texto, numero }

/// Una pregunta del formulario dinámico.
class AtributoPregunta {
  const AtributoPregunta({
    required this.key,
    required this.label,
    required this.tipo,
    this.options = const [],
    this.obligatoria = false,
    this.placeholder,
    this.badgeLabel,
    this.permiteCustom = false,
    this.showIfKey,
    this.showIfEquals,
  });

  /// Identificador estable; es la llave dentro del JSON `atributos` que se
  /// manda al servidor. Nunca renombrar una key ya publicada.
  final String key;

  /// Lo que lee el usuario, tanto al publicar como en el detalle.
  final String label;

  final AtributoTipo tipo;

  /// Valores permitidos (solo selección y selección múltiple).
  final List<String> options;

  /// Si hay que exigirla antes de publicar. Hoy todas están en false; ver la
  /// nota en la config del servidor.
  final bool obligatoria;

  /// Pista para los campos de texto y número.
  final String? placeholder;

  /// Texto corto para el badge de la tarjeta del listado. Solo lo necesitan
  /// los booleanos: su [label] es una pregunta y el badge tiene que ser una
  /// afirmación.
  final String? badgeLabel;

  /// Solo en selección múltiple: el usuario puede agregar valores propios
  /// además de los de [options]. Se guardan en la misma lista, sin marca que
  /// los distinga — para quien lee el detalle son un chip más.
  final bool permiteCustom;

  /// La pregunta solo aplica si la respuesta de [showIfKey] es
  /// [showIfEquals]. Se usa para los campos que cuelgan de un switch
  /// ("¿tiene garantía?" → "¿cuánto dura?").
  final String? showIfKey;
  final Object? showIfEquals;

  /// ¿Esta pregunta debe mostrarse, dado lo respondido hasta ahora?
  bool aplicaCon(Map<String, dynamic> respuestas) =>
      showIfKey == null || respuestas[showIfKey] == showIfEquals;
}

/// Topes de los valores personalizados de una pregunta con [permiteCustom].
/// Seis chips llenan dos renglones en un teléfono angosto y veinte caracteres
/// es lo que cabe en un chip sin truncarse. El servidor aplica los mismos.
const int maxOpcionesCustom = 6;
const int maxLargoOpcionCustom = 20;

/// Preguntas que aplican a TODAS las categorías.
const List<AtributoPregunta> atributosGenerales = [
  AtributoPregunta(
    key: 'acepta_devoluciones',
    label: '¿Aceptas devoluciones/reembolsos?',
    tipo: AtributoTipo.booleano,
  ),
  AtributoPregunta(
    key: 'precio_negociable',
    label: '¿El precio es negociable?',
    tipo: AtributoTipo.booleano,
  ),
  AtributoPregunta(
    key: 'lugar_entrega',
    label: '¿Entregas en punto de encuentro o solo en campus?',
    tipo: AtributoTipo.seleccion,
    options: ['Campus', 'Fuera del campus', 'Ambos'],
  ),
  AtributoPregunta(
    key: 'tiene_garantia',
    label: '¿Tiene garantía?',
    tipo: AtributoTipo.booleano,
  ),
  AtributoPregunta(
    key: 'duracion_garantia',
    label: '¿Cuánto dura la garantía?',
    tipo: AtributoTipo.texto,
    placeholder: 'Ej. 3 meses',
    showIfKey: 'tiene_garantia',
    showIfEquals: true,
  ),
];

/// Preguntas generales que NO aplican a ciertas categorías.
///
/// Las generales son generales porque le sirven a casi todo el catálogo, no
/// porque le sirvan a todo. Preguntarle a quien renta un cuarto si entrega en
/// el campus, o a quien vende tacos si acepta devoluciones, no es una pregunta
/// opcional de más: es una que no tiene respuesta posible, y quien la lee
/// entiende que el formulario no es para él.
///
/// Se excluye por key; los condicionales que cuelgan de una key excluida se
/// caen solos (ver [preguntasDeCategoria]).
const Map<String, List<String>> generalesExcluidas = {
  // La comida no se devuelve ni se garantiza, y su entrega ya la resuelve su
  // propia pregunta (`tipo_entrega`).
  'food': [
    'acepta_devoluciones',
    'precio_negociable',
    'lugar_entrega',
    'tiene_garantia',
  ],
  // Una renta no se entrega en un punto de encuentro: se visita.
  'housing': ['lugar_entrega'],
  // Unos apuntes no traen garantía; lo demás sí aplica.
  'notes': ['tiene_garantia'],
  // Un servicio no se devuelve ni se entrega en un lugar: para dónde se da
  // está `modalidad`.
  'services': ['acepta_devoluciones', 'lugar_entrega'],
};

/// Preguntas propias de cada categoría, indexadas por `MarketplaceCategory.id`.
const Map<String, List<AtributoPregunta>> atributosPorCategoria = {
  'books': [
    AtributoPregunta(
      key: 'edicion',
      label: '¿Es edición original o copia?',
      tipo: AtributoTipo.seleccion,
      options: ['Original', 'Copia/Fotocopia'],
    ),
    AtributoPregunta(
      key: 'estado_libro',
      label: 'Estado del libro',
      tipo: AtributoTipo.seleccion,
      options: ['Nuevo', 'Como nuevo', 'Buen estado', 'Con detalles/subrayado'],
    ),
  ],
  'clothes': [
    AtributoPregunta(
      key: 'talla',
      label: 'Talla',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. M, 28, 7½',
    ),
    AtributoPregunta(
      key: 'marca',
      label: 'Marca',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. Nike, Zara, sin marca',
    ),
    AtributoPregunta(
      key: 'condicion_etiqueta',
      label: '¿Conserva la etiqueta?',
      tipo: AtributoTipo.seleccion,
      options: ['Nuevo con etiqueta', 'Usado'],
    ),
    AtributoPregunta(
      key: 'estado_ropa',
      label: 'Estado de la prenda',
      tipo: AtributoTipo.seleccion,
      options: ['Nueva', 'Poco uso', 'Uso normal', 'Con detalles'],
    ),
    AtributoPregunta(
      key: 'cambio_talla',
      label: '¿Aplica cambio de talla si no queda?',
      tipo: AtributoTipo.booleano,
    ),
  ],
  'electronics': [
    AtributoPregunta(
      key: 'estado_electronico',
      label: 'Estado del equipo',
      tipo: AtributoTipo.seleccion,
      options: ['Nuevo', 'Seminuevo', 'Usado', 'Para refacciones'],
    ),
    AtributoPregunta(
      key: 'incluye_accesorios',
      label: '¿Incluye caja y accesorios originales?',
      tipo: AtributoTipo.booleano,
    ),
    AtributoPregunta(
      key: 'garantia_vigente',
      label: '¿Tiene garantía vigente?',
      tipo: AtributoTipo.booleano,
    ),
    AtributoPregunta(
      key: 'con_quien_garantia',
      label: '¿Con quién es la garantía?',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. Apple, Best Buy, la tienda donde lo compré',
      showIfKey: 'garantia_vigente',
      showIfEquals: true,
    ),
    AtributoPregunta(
      key: 'tiene_desperfecto',
      label: '¿Presenta algún desperfecto o detalle?',
      tipo: AtributoTipo.booleano,
    ),
    AtributoPregunta(
      key: 'descripcion_desperfecto',
      label: '¿Cuál es el desperfecto?',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. rayón en la esquina, la batería dura poco',
      showIfKey: 'tiene_desperfecto',
      showIfEquals: true,
    ),
    AtributoPregunta(
      key: 'tiempo_uso',
      label: 'Tiempo de uso aproximado',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 8 meses',
    ),
  ],
  'food': [
    AtributoPregunta(
      key: 'opciones_veg',
      label: '¿Tienes opciones vegetarianas/veganas?',
      tipo: AtributoTipo.booleano,
      badgeLabel: 'Opción veggie',
    ),
    AtributoPregunta(
      key: 'tipo_entrega',
      label: '¿Cómo entregas el pedido?',
      tipo: AtributoTipo.seleccion,
      options: ['Entrega a domicilio', 'Solo pickup', 'Ambos'],
    ),
    AtributoPregunta(
      key: 'personalizable',
      label: '¿Se puede personalizar el pedido?',
      tipo: AtributoTipo.booleano,
    ),
  ],
  'housing': [
    AtributoPregunta(
      key: 'incluye_renta',
      label: '¿Qué incluye la renta?',
      tipo: AtributoTipo.seleccionMultiple,
      options: ['Luz', 'Agua', 'Internet', 'Gas', 'Amueblado'],
      // Ningún catálogo cubre lo que incluye un cuarto ("Wifi 300mb",
      // "Limpieza semanal"), y alargar la lista tampoco: se vuelve una reja
      // de chips que nadie lee.
      permiteCustom: true,
    ),
    AtributoPregunta(
      key: 'requisitos',
      label: '¿Qué requisitos pides?',
      tipo: AtributoTipo.seleccionMultiple,
      options: ['Aval', 'Depósito', 'Contrato', 'Identificación'],
      permiteCustom: true,
    ),
    AtributoPregunta(
      key: 'acepta_mascotas',
      label: '¿Aceptas mascotas?',
      tipo: AtributoTipo.booleano,
      badgeLabel: 'Acepta mascotas',
    ),
    AtributoPregunta(
      key: 'exclusivo_para',
      label: '¿Hay alguna restricción de inquilino?',
      tipo: AtributoTipo.seleccion,
      options: [
        'Sin restricción',
        'Solo estudiantes',
        'Solo mujeres',
        'Solo hombres',
      ],
    ),
    AtributoPregunta(
      key: 'duracion_minima',
      label: 'Duración mínima de contrato',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 6 meses',
    ),
    AtributoPregunta(
      key: 'monto_deposito',
      label: '¿Cuánto es el depósito?',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. \$3,000 o un mes de renta',
    ),
    AtributoPregunta(
      key: 'distancia_universidad',
      label: 'Distancia a la universidad',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 10 min caminando',
    ),
  ],
  'notes': [
    AtributoPregunta(
      key: 'materia_profesor',
      label: 'Materia y profesor',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. Cálculo II — Ing. Ramírez',
    ),
    AtributoPregunta(
      key: 'periodo',
      label: 'Periodo/semestre',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. Otoño 2025',
    ),
    AtributoPregunta(
      key: 'formato',
      label: 'Formato',
      tipo: AtributoTipo.seleccion,
      options: ['Digital PDF', 'Físico', 'Ambos'],
    ),
    AtributoPregunta(
      key: 'completos',
      label: '¿Los apuntes están completos?',
      tipo: AtributoTipo.seleccion,
      options: ['Completos', 'Parciales'],
    ),
    AtributoPregunta(
      key: 'calificacion',
      label: 'Calificación con la que se aprobó',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 95',
    ),
  ],
  'services': [
    AtributoPregunta(
      key: 'tiempo_entrega',
      label: 'Tiempo estimado de entrega',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 3 días hábiles',
    ),
    AtributoPregunta(
      key: 'revisiones_incluidas',
      label: '¿Cuántas revisiones/ajustes incluidos?',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 2 revisiones sin costo',
    ),
    AtributoPregunta(
      key: 'modalidad',
      label: 'Modalidad',
      tipo: AtributoTipo.seleccion,
      options: ['Presencial', 'Remoto', 'Ambos'],
    ),
    AtributoPregunta(
      key: 'requiere_anticipo',
      label: '¿Requieres anticipo?',
      tipo: AtributoTipo.booleano,
    ),
    AtributoPregunta(
      key: 'porcentaje_anticipo',
      label: '¿De cuánto es el anticipo?',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. 50%',
      showIfKey: 'requiere_anticipo',
      showIfEquals: true,
    ),
    AtributoPregunta(
      key: 'tiene_portafolio',
      label: '¿Experiencia/portafolio disponible?',
      tipo: AtributoTipo.booleano,
    ),
  ],
  'other': [
    AtributoPregunta(
      key: 'estado_general',
      label: 'Estado general',
      tipo: AtributoTipo.texto,
      placeholder: 'Ej. Usado, funciona bien',
    ),
    AtributoPregunta(
      key: 'motivo_venta',
      label: 'Motivo de venta',
      tipo: AtributoTipo.texto,
      placeholder: 'Opcional',
    ),
    AtributoPregunta(
      key: 'razon_urgencia',
      label: '¿Por qué lo vendes con urgencia?',
      tipo: AtributoTipo.texto,
      placeholder: 'Opcional',
    ),
    AtributoPregunta(
      key: 'incluye_accesorios_otros',
      label: '¿Incluye accesorios?',
      tipo: AtributoTipo.booleano,
    ),
  ],
};

/// Todas las preguntas que aplican a una categoría: las generales que no estén
/// excluidas primero, luego las propias. Una categoría desconocida recibe solo
/// las generales en vez de reventar — un producto viejo con una categoría
/// retirada del catálogo debe seguir pudiéndose editar.
List<AtributoPregunta> preguntasDeCategoria(String categoryId) {
  final excluidas = {...?generalesExcluidas[categoryId]};
  final generales = <AtributoPregunta>[];
  for (final pregunta in atributosGenerales) {
    // Un hijo se va con su padre: el showIf de una pregunta que ya no se
    // pinta no se cumpliría nunca, y dejarla suelta la volvería visible.
    if (excluidas.contains(pregunta.key)) continue;
    if (pregunta.showIfKey != null && excluidas.contains(pregunta.showIfKey)) {
      excluidas.add(pregunta.key);
      continue;
    }
    generales.add(pregunta);
  }
  return [...generales, ...?atributosPorCategoria[categoryId]];
}

/// Busca una pregunta por su key dentro de una categoría. Devuelve null si no
/// existe: el detalle de un producto puede traer respuestas de una versión de
/// la app más nueva que esta, y esas simplemente no se pintan.
AtributoPregunta? preguntaPorKey(String categoryId, String key) {
  for (final pregunta in preguntasDeCategoria(categoryId)) {
    if (pregunta.key == key) return pregunta;
  }
  return null;
}

/// Deja solo las respuestas que siguen aplicando: descarta las de otra
/// categoría y los campos condicionales cuyo padre ya no se cumple.
///
/// Se aplica antes de mandar al servidor, para que apagar un switch borre de
/// verdad el texto que colgaba de él en vez de mandarlo y que el servidor lo
/// tenga que tirar. También se usa al cambiar de categoría en el formulario.
Map<String, dynamic> depurarRespuestas(
  Map<String, dynamic> respuestas,
  String categoryId,
) {
  final limpio = <String, dynamic>{};
  for (final pregunta in preguntasDeCategoria(categoryId)) {
    // El padre de un condicional siempre va declarado antes que el hijo, así
    // que `limpio` ya tiene su valor cuando toca evaluar el showIf.
    if (!pregunta.aplicaCon(limpio)) continue;

    final valor = respuestas[pregunta.key];
    if (valor == null) continue;
    if (valor is String && valor.trim().isEmpty) continue;
    if (valor is List && valor.isEmpty) continue;

    limpio[pregunta.key] = valor is String ? valor.trim() : valor;
  }
  return limpio;
}

File: backend/src/validation/atributosCategoria.js
const {
  preguntasDeCategoria,
  ATRIBUTOS_DESTACADOS,
  MAX_ATRIBUTOS_DESTACADOS,
} = require('../config/atributosCategoria');

/** Tope de caracteres de una respuesta de texto libre. */
const MAX_LARGO_TEXTO = 200;

/**
 * Normaliza y valida las respuestas a las preguntas dinámicas de una
 * categoría.
 *
 * Devuelve `{ value }` con un objeto listo para guardar, o `{ error }` con un
 * mensaje en español para el usuario. `value` es `null` cuando no quedó
 * ninguna respuesta: null y `{}` significan lo mismo para el cliente, y null
 * ahorra guardar una cadena `'{}'` en cada producto que no contestó nada.
 *
 * El criterio general es **descartar en silencio, rechazar solo lo que
 * miente**. Una key desconocida se ignora (un cliente viejo mandando un
 * atributo que ya se quitó no debe impedir publicar), pero un valor fuera de
 * las `options` declaradas sí es un error: significa que el cliente y el
 * servidor no coinciden en el catálogo, y guardarlo dejaría un producto con
 * un dato que la UI no sabe pintar.
 */
function validarAtributosCategoria(atributos, categoryId) {
  if (atributos === undefined || atributos === null || atributos === '') {
    return { value: null };
  }

  let parsed = atributos;
  if (typeof parsed === 'string') {
    // multipart/form-data manda todo como texto, así que el objeto llega
    // stringificado en las publicaciones con fotos.
    try {
      parsed = JSON.parse(parsed);
    } catch {
      return { error: 'atributos debe ser un JSON válido' };
    }
  }
  if (parsed === null) return { value: null };
  if (typeof parsed !== 'object' || Array.isArray(parsed)) {
    return { error: 'atributos debe ser un objeto de respuestas' };
  }

  const preguntas = preguntasDeCategoria(categoryId);
  const limpio = {};

  for (const pregunta of preguntas) {
    // El padre de un condicional siempre se resuelve antes que el hijo
    // porque en la config va declarado antes; así `limpio` ya tiene su valor
    // cuando toca evaluar el `showIf`.
    if (pregunta.showIf && limpio[pregunta.showIf.key] !== pregunta.showIf.equals) {
      continue;
    }

    const crudo = parsed[pregunta.key];
    const resultado = normalizarRespuesta(crudo, pregunta);
    if (resultado.error) return resultado;

    if (resultado.value === null) {
      if (pregunta.required) {
        return {
          error: `Responde "${pregunta.label}" para publicar.`,
          campo: pregunta.key,
        };
      }
      continue;
    }
    limpio[pregunta.key] = resultado.value;
  }

  return { value: Object.keys(limpio).length > 0 ? limpio : null };
}

/**
 * Normaliza UNA respuesta según el tipo de su pregunta. `{ value: null }`
 * significa "sin responder" (no es un error por sí solo; quien decide es el
 * `required` de la pregunta).
 */
function normalizarRespuesta(crudo, pregunta) {
  if (crudo === undefined || crudo === null || crudo === '') {
    return { value: null };
  }

  switch (pregunta.type) {
    case 'boolean': {
      // Los booleanos llegan como texto por multipart; 'false' es una cadena
      // no vacía y sería `true` con una conversión ingenua.
      if (typeof crudo === 'boolean') return { value: crudo };
      if (crudo === 'true') return { value: true };
      if (crudo === 'false') return { value: false };
      return { error: `"${pregunta.label}" debe responderse con sí o no.` };
    }

    case 'number': {
      const n = Number(crudo);
      if (!Number.isFinite(n)) {
        return { error: `"${pregunta.label}" debe ser un número.` };
      }
      return { value: n };
    }

    case 'select': {
      if (typeof crudo !== 'string' || !pregunta.options.includes(crudo)) {
        return { error: `Opción inválida en "${pregunta.label}": ${crudo}` };
      }
      return { value: crudo };
    }

    case 'multiselect': {
      let lista = crudo;
      if (typeof lista === 'string') {
        try {
          lista = JSON.parse(lista);
        } catch {
          return { error: `"${pregunta.label}" debe ser una lista.` };
        }
      }
      if (!Array.isArray(lista)) {
        return { error: `"${pregunta.label}" debe ser una lista.` };
      }
      const unicos = [...new Set(lista)];
      for (const opcion of unicos) {
        if (typeof opcion !== 'string' || !pregunta.options.includes(opcion)) {
          return { error: `Opción inválida en "${pregunta.label}": ${opcion}` };
        }
      }
      // Se reordenan según la config y no según lo que mandó el cliente, para
      // que dos productos con lo mismo incluido se lean igual en el detalle.
      const ordenados = pregunta.options.filter(o => unicos.includes(o));
      return { value: ordenados.length > 0 ? ordenados : null };
    }

    case 'text':
    default: {
      const texto = String(crudo).trim();
      if (!texto) return { value: null };
      if (texto.length > MAX_LARGO_TEXTO) {
        return {
          error: `"${pregunta.label}" no puede pasar de ${MAX_LARGO_TEXTO} caracteres.`,
        };
      }
      return { value: texto };
    }
  }
}

/**
 * Los 1-2 atributos que la tarjeta del listado muestra como badge, ya
 * resueltos a `{ key, label, value }` para que el cliente no tenga que
 * consultar el catálogo de preguntas al pintar una cuadrícula.
 *
 * Se calcula en el servidor y no en Flutter porque la tarjeta se pinta en
 * cuatro pantallas distintas (home, búsqueda, perfil de vendedor, carruseles
 * del detalle) y ninguna debería tener su propia opinión sobre cuál es el
 * dato importante de un producto de ropa.
 *
 * Los booleanos solo aparecen si son `true`: un badge "No acepta mascotas" en
 * una cuadrícula es ruido, y su ausencia no afirma lo contrario — para eso
 * está el detalle.
 */
function atributosDestacados(atributos, categoryId) {
  if (!atributos || typeof atributos !== 'object') return [];

  const keys = ATRIBUTOS_DESTACADOS[categoryId] || [];
  if (keys.length === 0) return [];

  const preguntas = preguntasDeCategoria(categoryId);
  const destacados = [];

  for (const key of keys) {
    if (destacados.length >= MAX_ATRIBUTOS_DESTACADOS) break;

    const pregunta = preguntas.find(p => p.key === key);
    const valor = atributos[key];
    if (!pregunta || valor === undefined || valor === null || valor === '') continue;

    if (pregunta.type === 'boolean') {
      if (valor !== true) continue;
      destacados.push({
        key,
        label: pregunta.label,
        value: pregunta.badgeLabel || pregunta.label,
      });
      continue;
    }
    if (Array.isArray(valor)) {
      if (valor.length === 0) continue;
      destacados.push({ key, label: pregunta.label, value: valor.join(' · ') });
      continue;
    }
    destacados.push({ key, label: pregunta.label, value: String(valor) });
  }

  return destacados;
}

module.exports = {
  validarAtributosCategoria,
  atributosDestacados,
  MAX_LARGO_TEXTO,
};
